testssl.sh

```diff
@@ -17379,51 +17373,92 @@ sclient_auth() {
 #
 determine_optimal_proto_sockets_helper() {
      local all_failed=true
-     local proto=""
+     local proto="" p
      local optimal_proto=""
      local starttls="$1"
+     local -i ret
 
      for proto in 03 01 04 00 02 22; do
           case $proto in
                03) tls_sockets "$proto" "$TLS12_CIPHER"
-                    if [[ $? -eq 0 ]]; then
+                    ret=$?
+                    if [[ $ret -eq 0 ]]; then
                          add_tls_offered tls1_2 yes; optimal_proto="-tls1_2"
                          all_failed=false
-                         break
-                    elif [[ $? -eq 2 ]]; then
+                    elif [[ $ret -eq 2 ]]; then
                          case $(get_protocol "$TEMPDIR/$NODEIP.parse_tls_serverhello.txt") in
                               *1.1)   add_tls_offered tls1_1 yes; optimal_proto="-tls1_1" ;;
                               TLSv1)  add_tls_offered tls1 yes; optimal_proto="-tls1" ;;
                               SSLv3)  add_tls_offered ssl3 yes; optimal_proto="-ssl3" ;;
                          esac
                          all_failed=false
-                         break
-                    fi ;;
+                    fi
+                    # Try again with a different, less common, set of cipher suites
+                    # see #807 and #806
+                    if [[ $ret -ne 0 ]]; then
+                         tls_sockets "$proto" "$TLS12_CIPHER_2ND_TRY"
+                         ret=$?
+                         if [[ $ret -eq 0 ]]; then
+                              add_tls_offered tls1_2 yes; optimal_proto="-tls1_2"
+                              TLS12_CIPHER="$TLS12_CIPHER_2ND_TRY"
+                              all_failed=false
+                         else
+                              add_tls_offered tls1_2 no
+                         fi
+                         if [[ $ret -eq 2 ]]; then
+                              case $(get_protocol "$TEMPDIR/$NODEIP.parse_tls_serverhello.txt") in
+                                   *1.1)   p="tls1_1" ;;
+                                   TLSv1)  p="tls1" ;;
+                                   SSLv3)  p="ssl3" ;;
+                              esac
+                              add_tls_offered "$p" yes
+                              if [[ -z "$optimal_proto" ]]; then
+                                   TLS12_CIPHER="$TLS12_CIPHER_2ND_TRY"
+                                   optimal_proto="-${p}"
+                              fi
+                              all_failed=false
+                         fi
+                    fi
+                    "$all_failed" || break ;;
                04) tls_sockets "$proto" "$TLS13_CIPHER"
                     if [[ $? -eq 0 ]]; then
                          add_tls_offered tls1_3 yes; optimal_proto="-tls1_3"
                          all_failed=false
-                         break
+                         # Don't break if this test succeeds, as some tests
```

Comment:
Don't want to do a full protocol check here. For this task run_protocols() should be used.

determine_optimal_proto() is ONLY for dealing with broken servers which otherwise -- no protocol supplied in a openssl clienthello -- refuse serve their expected serverhello properly. See comment above determine_optimal_proto().